File: npr.pl
#!/usr/bin/perl

use MIME::Lite;
use File::Copy;
#use Cal::Date qw(today ISO_day MJD);
@months = ('Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec');

#%programs = (WC=>39, atc=>2, ME=>3, TOTN=>5, WESAT=>7, WESUN=>10, FA=>13, DAY=>17, WW=>35);
#%programs = (WC=>39, atc=>'all_things_considered', ME=>'morning_edition', TOTN=>5, WESAT=>7, WESUN=>10, FA=>13, DAY=>17, WW=>35);
%programs = (wc=>39, atc=>'all-things-considered', me=>'morning-edition', totn=>5, wesat=>'weekend-edition-saturday', wesun=>'weekend-edition-sunday', fa=>'fresh-air', ww=>'wait-wait-dont-tell-me');

$prog = $ARGV[0];
chomp($prog);
$dir = "/var/www/html";

$pgid = $programs{$prog};
if(@ARGV == 1){
# prepare datestring for current day
	chomp($wdate = `date +%w`);
	chomp($tod = `date +%Y%m%d`);
# 20050429
	$year = substr($tod, 0,4);
	$day = substr($tod, -2,2);
	$month = substr($tod, -4,2);
	$mo = $months[(substr($tod, -4,2))-1];
	$dstr = "$tod";
	$urldate = "$year-$month-$day";
	$mark = "/var/tmp/" . $prog . $month . $day;
	if(-e "$mark"){
		&logg("Flag file $mark exists, exiting");
		exit();
	}	
	$arch="http://www.npr.org/programs/$pgid/archive";
## need error checking on this wget...
	chomp($preurl = `wget -qO - $arch |grep 'showDate=$urldate' |head -1`);
	if($? != 0){
		&domail("Error: request failed for $urldate archive page for $pgid.");
		exit(2);
	}	
#	chomp($preurl = `wget -O - $arch |grep 'showDate=$urldate\">'`);
	if($preurl =~ /href=\"([^\"]+)\"/){
		$url = $1;
	}
	else{
		print "Error: URL for $urldate not found in archive page for $pgid.\n";
		print "Pre-URL contains: $preurl\n";
		exit(1);
	}
	if(-e "$dir/$prog.m3u"){
		system("cat $dir/$prog.m3u >>$dir/$prog.log");
		chomp($wc = `wc -l $dir/$prog.m3u`);
		copy("$dir/$prog.m3u", "/var/tmp/");
	}
	open(M3U, ">$dir/$prog.m3u") or die "Can't open file:$!\n";
	copy("$dir/$prog.m3u", "$dir/$prog.m3u.last");
#	open(wM3U, ">$dir/wm-$prog.m3u") or die "Can't open file:$!\n";
#	$flag = "/var/tmp/" . $prog;
	$flag = "/var/tmp/" . $prog . $month . $day;
}
# datestring can be from any date in the past
elsif(@ARGV == 2) {
	chomp($dat = $ARGV[1]);
#	$pgid = $programs{$pcode};
	print "pgid = $pgid\n";
	$day = substr($dat, -2,2);
	$month = substr($dat, -4,2);
	$mo = $months[$month-1];
	$year = substr($dat, 0,4);
	$fdstr = substr($dat, -4,4); 
	$dstr = "$dat";
	$urldate = "$year-$month-$day";
	#$url = "http://www.npr.org/programs/$prog";
	$arch="http://www.npr.org/programs/$pgid/archive";
#	<a class="full-show" href="http://www.npr.org/programs/morning-edition/2013/12/20/255698889/morning-edition-for-december-20-2013?showDate=2013-12-20
#	<a href="http://www.npr.org/programs/morning-edition/2014/03/21/292225680/morning-edition-for-march-21-2014?showDate=2014-03-21"  data-metrics='{"action":"Click Archive View Full Show","category":"News Magazine Archive","label":"http:\/\/www.npr.org\/programs\/morning-edition\/2014\/03\/21\/292225680\/morning-edition-for-march-21-2014"}' >
	chomp($preurl = `wget -qO - $arch |grep 'showDate=$urldate' |head -1`);
#	chomp($preurl = `wget -O - $arch |grep 'showDate=$urldate\">'`);
	if($preurl =~ /href=\"([^\"]+)\"/){
		$url = $1;
	}
	else{
		print "Error: URL for $urldate not found in archive page for $pgid.\n";
		print "Debug: $preurl\n";
		exit(1);
	}
	#$url = "http://www.npr.org/templates/rundowns/rundown.php?prgId=$pgid\&prgDate=$urldate\&view=storyview";
	$flag = "/var/tmp/" . $prog . $month . $day;
### if grabbing archive, don't overwrite today's playlist, make new
	open(M3U, ">/var/www/html/$prog-$fdstr.m3u") or die "Can't open file:$!\n";
}
else {
	print "Usage: $0 <profile> [<MMDDYYYY>].\n";
	print " Date defaults to current day.\n";
	exit(1);
}

if($pcode eq "ww"){
	$pcode = "waitwait";
}

$blat = "/var/www/html/nprout";

if($ENV{TERM}){
	print "Requesting $url...\n";
}
system("wget -qO $blat '$url'");
#system("perl -p -i -e \'s/></>\n</g\' $blat");
open(OUTPUT, $blat); # should test for age
#$cnt = 1;
my @plist;
foreach $line (<OUTPUT>){
#	<input type="hidden" id="title252006903" value="Obama Meets With Tech Leaders, Taps Microsoft Exec To Fix HealthCare.gov"></input>
#<h3 class="rundown-segment__title"><a href="http://www.npr.org/2016/12/24/506849921/how-kitchen-sounds-influence-food-flavor"  data-metrics='{"action":"Click Rundown Story 5","category":"Rundown Click","label":"http:\/\/www.npr.org\/2016\/12\/24\/506849921\/how-kitchen-sounds-influence-food-flavor"}' >How Kitchen Sounds Influence Food Flavor</a></h3>
#
	chomp;
#	if($line =~ /id=\"title([0-9]+)\" value=\"([^\"]+)\"/){
	if($line =~ /' >([^<]+)<\/a><\/h3>/){
		$tit = "unknown title";
#		chomp($sid = $1);
#		chomp($tit = $2);
		chomp($tit = $1);
		next if($tit =~ /.+$day, $year$/);
#		$tit = $2 if($2 =~ /\w/);
		$tit =~ s/[\(\*\)]//g;
		$tit =~ s/[\(\;\)]//g;
		$tit =~ s/[\@\?\']//g;
		$tit =~ s/ To / to /g;
		$tit =~ s/ The / the /g;
		$tit =~ s/ Or / or /g;
		$tit =~ s/ And / and /g;
		$tit =~ s/ With / with /g;
		$tit =~ s/ For / for /g;
		print "Extracted title: \'$tit\'\n" if($ENV{TERM});
		if(@plist > 1){
			for($i=0; $i<=$#plist; $i++){
				if($plist[$i] =~ /$tit/){
					splice(@plist, $i, 2);
				} 
			}
		}
		push(@plist, "\# $tit");
		$pcode = lc($pcode);
#		http://pd.npr.org/anon.npr-mp3/npr/me/2013/12/20131218_me_24.mp3
#		$mp3url = "http://pd.npr.org/anon.npr-mp3/npr/$pcode/$year/$month/$dstr" . "_" . $pcode . "_";
#  		push(@plist, $mp3url);
#		push(@plist, "http://www.npr.org/dmg/dmg.php\?prgCode=$pcode\&showDate=$dstr\&segNum=");
		$tt++;
#		$cnt++;
	}
#	<li class="download"><a href="http://pd.npr.org/anon.npr-mp3/npr/me/2013/12/20131218_me_opponents_to_challenge_calif_school_act.mp3?dl=1"  data-metrics=
#	 <li><a class="audio-tool audio-tool-download" href="http://pd.npr.org/anon.npr-mp3/npr/atc/2016/03/20160329_atc_remembering_patty_duke_hollywoods_miracle_worker.mp3?orgId=1&amp;topicId=1062&amp;d=251&amp;p=2&amp;story=472309592&amp;t=progseg&amp;e=472228928&amp;seg=19&amp;siteplayer=true&amp;dl=1" 
#<a href="http://www.npr.org/2016/04/12/473992245/criminal-justice-dominates-crowded-baltimore-mayoral-race"  data-metrics='{"action":"Click Rundown Story 1","category":"Rundown Click","label":"http:\/\/www.npr.org\/2016\/04\/12\/473992245\/criminal-justice-dominates-crowded-baltimore-mayoral-race"}' >Criminal Justice Dominates Crowded Baltimore Mayoral Race</a></h1>
# http://pd.npr.org/anon.npr-mp3/npr/atc/2016/04/20160412_atc_criminal_justice_dominates_crowded_baltimore_mayoral_race.mp3?orgId=1&topicId=1014&d=235&p=2&story=473992245&t=progseg&e=473914722&seg=1&siteplayer=true&dl=1

#audio-tool-download"><a href
#
	elsif($line =~ /download\"><a href=\"([^\"]+\.mp3).+/){
		$mp3url = $1;
		next if($mp3url !~ /$dstr/);
#		$mp3url =~ s/\?dl=1$//;
		push(@plist, $mp3url);
		print "Extracted URL: \'$mp3url\'\n" if($ENV{TERM});
		$tx++;
	}
}
close(OUTPUT);
# if no matches, audio links are not ready yet
if(@plist < 1) {
	print "The regex match is failing for $prog.\n" if($ENV{TERM});
#	&domail("The regex match is failing for $prog.");
	exit(1);
}

if(($prog eq "me")&&($tx < 10)&&(!$ENV{TERM})){
	&logg("Found too few story links for $prog ($tx), discarding playlist");
#	unlink("$dir/$prog.m3u") or &logg("unlink failed on $dir/$prog.m3u");
#	unlink($flag) or &logg("unlink failed on $flag");
	$playlist = $dir . "/" . $prog . ".m3u";
	$rmp = "/bin/rm " . $playlist;
	if(system($rmp)!=0){
		&logg("rm failed on $playlist");
	}
	else{
		&logg("removed file $playlist");
		if(-e "$playlist"){
			&logg("but it's still there");
		}
	}
#	$rmf = "/bin/rm " . $flag;
#	if(system($rmf)!=0){
#		$r = $?;
#		&logg("rm failed on $flag ($r)");
#	}
#	if(-e "$flag"){
#		unlink($flag);
#	}
	exit();
}
if(($wdate =~ /[1-5]/)&&($prog eq "atc")&&($tx < 12)&&(!$ENV{TERM})){
	&logg("Found too few story links for $prog ($tx), discarding playlist");
#	unlink("$dir/$prog.m3u") or &logg("unlink failed on $dir/$prog.m3u");
#	unlink($flag) or &logg("unlink failed on $flag");
	$playlist = $dir . "/" . $prog . ".m3u";
	$rmp = "/bin/rm " . $playlist;
	if(system($rmp)!=0){
		&logg("rm failed on $playlist");
	}
	else{
		&logg("removed file $playlist");
		if(-e "$playlist"){
			&logg("but it's still there");
		}
	}
#	if(-e "$flag"){
#		unlink($flag);
#	}
	exit();
}
if(($prog =~ /^we/)&&($tx < 14)){
#	&domail("Found too few story links for $prog ($tx), discarding playlist");
	&logg("Found too few story links for $prog ($tx), discarding playlist");
	$playlist = $dir . "/" . $prog . ".m3u";
	$rmp = "/bin/rm " . $playlist;
	if(system($rmp)!=0){
		&logg("rm failed on $playlist ($?)");
	}
	else{
		&logg("removed file $playlist");
		if(-e "$playlist"){
			&logg("but it's still there");
		}
	}
#	$rmf = "/bin/rm " . $flag;
#	if(system($rmf)!=0){
#		$r = $?;
#		&logg("unlink failed on $flag ($r)");
#	}
#	if(-e "$flag"){
#		unlink($flag);
#	}
	exit();
}
# do the actual work of building the file
$cnt=0;
foreach $line (@plist){
	if($line =~ /^http/){
		$cnt++;
		$tx = $cnt;
		if(($pcode eq "me")&&($cnt == (@plist/2 - 1))){
			$cnt = 50;
		}
		$num = $cnt;
		if($cnt < 10){
			$num = "0" . $cnt;
		}
#		$line .= "$num\.mp3";
#		$line .= "$num\.mp3?dl=1";
	}
	print M3U "$line\n";
#	print "$line\n";
#	$line =~ s/RM/WM/;
#	print wM3U "$line\n";
}
if($ENV{TERM}){
	print "Found $cnt story links for $prog\n";
}
close(M3U);
## this seems to incorrectly unlink $flag...
#if($wc){
#	chomp($nwc = `wc -l $dir/$prog.m3u`);
#	if($nwc > $wc){
#		unlink($flag) or warn "unlink failed on $flag: $!\n";
#	}
#}
if($ENV{TERM}){
	print "Finished reporting, $cnt links in $prog.m3u.\n";
	exit();
}
#&domail();

sub domail(){
	$tim = localtime(time);
	if(@_){
		$subject = "Error: too few links for $prog";
		$msg = "@_";
#		$subject = "Error: no match for $prog";
#		$msg = "The regex match is failing for $prog\n";
		&logg($msg);
	}
	else{
		$subject = "Playlist prepared for $prog";
#		$msg = "\nhttp://www.pyoing.net/$prog.m3u\n";
		$msg .= "Tracks: $tx\n\n";
	}
	$msg .= "Message generated $tim by $0\n";
	$msg = MIME::Lite->new(
		SetSender=>'rserling@comcast.net',
		Return-Path =>'rserling@comcast.net',
		From    =>'rserling@comcast.net',
#		To      =>'elyons@pyoing.net',
		To      =>'bitshag@gmail.com',
		#Cc      =>'bergie@bergie.net',
		Subject =>"$subject",
		Type    =>'text/plain',
		Data =>"$msg"
	);
	$msg->send;
#	if($msg->send){
#		exit();
#		print "send from $sender successful.\n";
#	}
#	else {
#		exit(1);
#		print "send failed.\n";
#	}
}
sub logg($){
	$msg = "@_";
	open(LOG, ">>/var/log/nprgrab");
	select LOG unless($ENV{TERM});
	chomp($d = `date +\"\%b \%d \%T\"`);
	print "[$d] (npr) $msg\n";
	close(LOG);
}

File: npr.py
#!/usr/bin/env python3

import json
import os
import re
import sys
import shutil
import subprocess
import time
from datetime import datetime, timedelta
from pathlib import Path
import smtplib
from email.mime.text import MIMEText

# Constants
PROGRAMS = {
    'atc': 'all-things-considered',
    'fa': 'fresh-air',
    'me': 'morning-edition',
    'wesat': 'weekend-edition-saturday',
    'wesun': 'weekend-edition-sunday',
    'ww': 'wait-wait-dont-tell-me'
}

TITLES = {
    'atc': 'All Things Considered',
    'fa': 'Fresh Aire',
    'me': 'Morning Edition',
    'wesat': 'Weakened Edition Saturday',
    'wesun': 'Weakened Edition Sunday',
    'ww': 'Wait Wait Dont Tell Me'
}

BASE = os.getenv("HOME")
LPATH = os.path.join(BASE, "Music")
NPATH = os.path.join(LPATH, "radio")
TPATH = "/var/tmp"

def log_message(message):
    timestamp = datetime.now().strftime("%b %d %H:%M:%S")
    log_file = os.path.join(LPATH, "nprgrab.log")
    log_entry = f"[{timestamp}] (npr) {message}\n"
    if 'TERM' in os.environ:
        print(log_entry, end='')
    with open(log_file, 'a') as f:
        f.write(log_entry)

def send_email(subject, body):
    hostname = subprocess.getoutput('hostname')
    timestamp = datetime.now().strftime("%c")
    body += f"\nMessage generated {timestamp} by {hostname}:{sys.argv[0]}\n"
    msg = MIMEText(body)
    msg['Subject'] = subject
    msg['From'] = os.getenv('SMTP_FROM', 'wolfstools@gmail.com')
    msg['To'] = 'bitshag@gmail.com'
    try:
        with smtplib.SMTP('localhost') as s:
            s.send_message(msg)
        return True
    except Exception as e:
        log_message(f"ERROR: Email send failed: {str(e)}")
        return False

def scrape(prog, prg, urldate, year, month, day, wdate, dir_path, fdstr=None):
    pgid = PROGRAMS[prg]
    arch = f"http://www.npr.org/programs/{pgid}/archive"

    try:
        cmd = f"wget -qO - {arch} | grep 'showDate={urldate}' | head -1"
        preurl = subprocess.check_output(cmd, shell=True).decode('utf-8').strip()
        if not preurl:
            log_message(f"ERROR: URL for {urldate} not found in archive page for {pgid}.")
            sys.exit(1)
        url_match = re.search(r'href="([^"]+)"', preurl)
        if not url_match:
            log_message(f"ERROR: Could not extract URL from archive page response")
            sys.exit(1)
        url = url_match.group(1)
    except subprocess.CalledProcessError:
        log_message(f"ERROR: request failed for {urldate} archive page for {pgid}.")
        sys.exit(2)

    playlist_path = f"{dir_path}/{prg}.m3u"
    if fdstr is None:
        if os.path.exists(playlist_path):
            with open(playlist_path, 'r') as f:
                sum(1 for _ in f)
            shutil.copy2(playlist_path, f"{playlist_path}.last")
        m3u_file = open(playlist_path, 'w')
    else:
        m3u_file = open(f"{dir_path}/{prg}-{fdstr}.m3u", 'w')

    output_file = "/var/tmp/nprout"
    log_message(f"requesting url {url} and writing to {output_file}")
    subprocess.run(['wget', '-qO', output_file, url])

    playlist = []
    with open(output_file, 'r') as f:
        content = f.read()
        tblob = re.findall(r"\" data-audio='([^\']+)'", content)
        size = len(tblob)
        log_message(f"DEBUG: Found {size} audioUrl matches")
        if size == 0:
            log_message(f"ERROR: tblob has no size, exiting")
            sys.exit(2)
        title = ""
        for story in tblob:
            try:
                data = json.loads(story)
            except (json.JSONDecodeError, Exception) as e:
                log_message(f"ERROR: JSON parse failed: {e}")
                continue
            if "title" not in data or "audioUrl" not in data:
                log_message(f"ERROR: Missing required keys in JSON")
                continue
            tit = data["title"]
            if tit == title:
                continue
            title = tit
            if prg == 'me':
                title_lower = title.lower()
                if 'news brief' in title_lower or 'newsbrief' in title_lower:
                    log_message("INFO: Skipping Morning News Brief")
                    continue
            playlist.append(f"# {title}")
            rawurl = data["audioUrl"]
            playlist.append(rawurl.split("?")[0])

    if not playlist:
        log_message(f"ERROR: playlist has no size, exiting")
        sys.exit(2)

    for item in playlist:
        m3u_file.write(f"{item}\n")
    m3u_file.close()

    tx = len([u for u in playlist if u.startswith('http')])
    if prg == 'me' and tx < 10 and not os.environ.get('TERM'):
        log_message(f"Found too few story links for {prg} ({tx}), discarding playlist")
        try:
            os.remove(playlist_path)
        except OSError as e:
            log_message(f"ERROR: Failed to remove playlist: {e}")

    if prg == 'atc' and wdate in '12345' and tx < 12 and not os.environ.get('TERM'):
        log_message(f"ERROR: Found too few story links for {prg} ({tx}), discarding playlist")
        try:
            os.remove(playlist_path)
        except OSError as e:
            log_message(f"ERROR: Failed to remove playlist: {e}")

def grab(prg, date_str, year, playlist_path):
    mark_file = os.path.join(TPATH, f"{prg}{date_str}")

    if os.path.exists(mark_file):
        log_message(f"WARN: job already completed for {prg}{date_str}")
        sys.exit(0)

    if not os.path.exists(playlist_path):
        log_message(f"ERROR: playlist file {playlist_path} is absent")
        sys.exit(0)

    if os.path.getsize(playlist_path) == 0:
        log_message(f"ERROR: playlist file {playlist_path} has no size, deleting")
        os.unlink(playlist_path)
        sys.exit(0)

    mtime = os.path.getmtime(playlist_path)
    if (time.time() - mtime) > 28800:
        log_message(f"WARN: playlist file {playlist_path} is stale, aborting")
        os.unlink(playlist_path)
        sys.exit(3)

    os.chdir(TPATH)
    done = 0

    with open(playlist_path) as f:
        story = "Unknown Title"
        track_num = 0
        for line in f:
            line = line.strip()
            if line.startswith('# '):
                new_story = line[2:]
                if new_story == story:
                    continue
                story = new_story.replace("'", "")
                for word in ['Or', 'Of', 'And', 'The', 'For', 'A', 'An']:
                    story = story.replace(f" {word} ", f" {word.lower()} ")
                continue
            if line.endswith('.mp3') and line.startswith('http'):
                track_num += 1
                track_str = f"{track_num:02d}"
                output_name = f"{prg}{date_str}t{track_str}.mp3"
                if os.path.exists(os.path.join(NPATH, output_name)):
                    log_message(f"WARN: File {output_name} already exists, skipping.")
                    done += 1
                    continue
                if 'TERM' in os.environ:
                    print(f"Grabbing {prg}{date_str} track {track_str}...")
                try:
                    subprocess.run(['wget', '-qO', output_name, line], check=True)
                    lame_cmd = [
                        'lame', '-a', '-m', 'm', '-b', '24', '--resample', '16',
                        '--scale', '2', '--quiet', '--add-id3v2',
                        '--ta', 'NPR',
                        '--tl', f"{date_str} {TITLES[prg]}",
                        '--ty', year,
                        '--tn', track_str,
                        '--tg', 'Speech'
                    ]
                    if story != "Unknown Title":
                        lame_cmd.extend(['--tt', story])
                    lame_cmd.extend([output_name, os.path.join(NPATH, output_name)])
                    subprocess.run(lame_cmd, check=True, capture_output=True)
                    done += 1
                    os.unlink(output_name)
                except subprocess.CalledProcessError as e:
                    log_message(f"ERROR: Error processing {output_name}: {str(e)}")

    if done < 1:
        log_message("ERROR: no files available after downloads")
        sys.exit(1)

    with open(mark_file, 'w') as f:
        f.write(str(done))
    files_text = "file" if done == 1 else "files"
    log_message(f"INFO: Re-encoded and tagged {done} {files_text} for {prg}")

def main():
    usage = (f"Usage: {sys.argv[0]} <program> [<days_ago>]\n"
             f" where program can be atc, fa, me, wesat, wesun, ww\n"
             f" optional second arg is number of days ago\n")

    if len(sys.argv) < 2:
        print(usage)
        sys.exit(2)

    prg = sys.argv[1].strip()
    if prg not in PROGRAMS:
        print(usage)
        print(f"({prg})")
        sys.exit(2)

    fdstr = None
    if len(sys.argv) > 2:
        try:
            days_ago = int(sys.argv[2])
            target_date = datetime.now() - timedelta(days=days_ago)
            fdstr = target_date.strftime("%m%d")
        except ValueError:
            print(usage)
            sys.exit(2)
    else:
        target_date = datetime.now()

    year  = target_date.strftime("%Y")
    month = target_date.strftime("%m")
    day   = target_date.strftime("%d")
    wdate = target_date.strftime("%w")
    date_str = f"{month}{day}"
    urldate = f"{year}-{month}-{day}"

    if fdstr is None:
        mark = os.path.join(TPATH, f"{prg}{date_str}")
        if os.path.exists(mark):
            log_message(f"WARN: Flag file {mark} exists, exiting")
            sys.exit(0)

    playlist_path = os.path.join(LPATH, f"{prg}.m3u")

    scrape(prg, prg, urldate, year, month, day, wdate, LPATH, fdstr)
    grab(prg, date_str, year, playlist_path)

if __name__ == "__main__":
    main()
